stan/math/prim/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,121 @@
+#ifndef STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_PRIM_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/inverse_spd.hpp>
+#include <stan/math/prim/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/mdivide_right_spd.hpp>
+#include <stan/math/prim/fun/inverse.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the Moore-Penrose generalized inverse of the specified matrix.
+ *
+ * The method is based on the Cholesky computation of the transform as specified
+ in
+ *
+ * <ul><li> Courrieu, Pierre. 2008.  Fast Computation of Moore-Penrose Inverse
+ Matrices.
+ * <i>arXiv</i> <b>0804.4809</b> </li></ul>
+ *
+ * @tparam EigMat type of the matrix (must be derived from \c Eigen::MatrixBase)
+ *
+ * @param G specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ * @note Because the method inverts a SPD matrix internally that interal matrix
+ may result in small numerical issues that result in a non-SPD error. There are
+ two
+ * <code>generalized_inverse</code> functions, one that uses one input matrix
+ (this one)
+ * and another that works with an input matrix and a small jitter to the
+ diagonal of the internal SPD
+ * matrix.
+ */
+template <typename EigMat, require_eigen_t<EigMat>* = nullptr>
+inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
+                     EigMat::ColsAtCompileTime>
+generalized_inverse(const EigMat& G) {
+  using value_t = value_type_t<EigMat>;
+  if (G.size() == 0) {
+    return {};
+  }
+
+  const auto n = G.rows();
+  const auto m = G.cols();
+
+  if (G.rows() == G.cols()) {
+    return inverse(G);
+  }
+
+  if (n < m) {
+    return transpose(mdivide_left_spd(tcrossprod(G), G));
+  } else {
+    return transpose(mdivide_right_spd(G, crossprod(G)));
+  }
```

Comment:
`EigMat` can potentially be an Eigen expression like `Product<Matrix, Matrix>` which can be bad to use multiple times because `G` will be evaluated each time it's called. When you use an eigen input multiple times you should use `auto&& G_ref = to_ref(G)`. `to_ref` looks at an Eigen type and if it's an expression will evaluate it. If `EigMat` is a plain type (i.e. like an Eigen::Matrix` then `to_ref` just passes the object back by reference. So for here you want like 

```suggestion
  if (G.rows() == G.cols()) {
    return inverse(G);
  }
  const auto& G_ref = to_ref(G);
  if (n < m) {
    return transpose(mdivide_left_spd(tcrossprod(G_ref), G_ref));
  } else {
    return transpose(mdivide_right_spd(G_ref, crossprod(G_ref)));
  }
```

Reply:
You only need to do it after the first if because if the `if` is true then `G` is only used once